sklearn/mixture/bayesian_mixture.py

```diff
@@ -0,0 +1,651 @@
+from .gaussian_mixture import _estimate_gaussian_parameters
+from .gaussian_mixture import _estimate_log_gaussian_prob
+from ..utils import check_array
+from ..utils.validation import check_is_fitted
+
+
+def _log_dirichlet_norm(alpha):
+    """Compute the log of the Dirichlet distribution normalization term.
+
+    Parameters
+    ----------
+    alpha : array-like, shape (n_samples,)
+        The parameters values of the Dirichlet distribution.
+
+    Returns
+    -------
+    log_dirichlet_norm : float
+        The log normalization of the Dirichlet distribution.
+    """
+    return gammaln(np.sum(alpha)) - np.sum(gammaln(alpha))
+
+
+def _log_wishart_norm(nu, log_det_precisions_chol, n_features):
+    """Compute the log of the Wishart distribution normalization term.
+
+    Parameters
+    ----------
+    nu : array-like, shape (n_components,)
+        The parameters values of the Whishart distribution.
+
+    log_det_precision_chol : array-like, shape (n_components,)
+        The determinant of the cholesky decomposition of the precision matrix.
+
+    n_features : int
+        The number of features.
+
+    Return
+    ------
+    log_wishart_norm : array-like, shape (n_components,)
+        The log normalization of the Wishart distribution.
+    """
+    # To simplify the computation we have removed the np.log(np.pi) term
+    return -(
+        nu * log_det_precisions_chol + nu * n_features * .5 * np.log(2.) +
+        np.sum(gammaln(.5 * (nu -
+                             np.arange(0, n_features)[:, np.newaxis])), 0))
+
+
+class BayesianGaussianMixture(BaseMixture):
+    """Variational estimation of a Gaussian mixture.
+
+    Representation of a variational inference for a Bayesian Gaussian mixture
+    model probability distribution. This class allows to do inference of an
+    approximate posterior distribution over the parameters of a Gaussian
+    mixture distribution. The number of components can be inferred from the
+    data.
+
+    Read more in the :ref:`User Guide <bgmm>`.
+
+    Parameters
+    ----------
+    n_components: int, defaults to 1.
+        The number of mixture components.
+
+    covariance_type : {'full', 'tied', 'diag', 'spherical'},
+        defaults to 'full'.
+        String describing the type of covariance parameters to use.
+        Must be one of::
+        'full' (each component has its own general covariance matrix).
+        'tied' (all components share the same general covariance matrix),
+        'diag' (each component has its own diagonal covariance matrix),
+        'spherical' (each component has its own single variance),
+
+    tol : float, defaults to 1e-3.
+        The convergence threshold. EM iterations will stop when the
+        log_likelihood average gain is below this threshold.
+
+    reg_covar : float, defaults to 0.
+        Non-negative regularization added to the diagonal of covariance.
+        Allows to assure that the covariance matrices are all positive.
```

Comment:
Do we really need this kind of hackish covariance regularization in the case of Bayesian mixture models? Aren't the bias imposed by the prior(s) enough to avoid degenerate covariance estimation?